### ISSUE TITLE
Converting assets in target directory instead of source one

### ISSUE THREAD

**Issue by DrDeath72** [yiisoft/yii2#10195]
Please consider converting assets in target directory instead of doing it in the source one. At least when `linkAssets` is turned on. Currently it leaves files in source dir and these files aren't deleted when assets are cleaned up. That's how I implemented it in my case. It's putting less file into `/assets/{hash}.css`.

``` php
<?

namespace app\components\web;

use Yii;

class AssetConverter extends \yii\web\AssetConverter {

    public function convert($asset, $basePath) {
        $am = Yii::$app->assetManager;
        if(!$am->linkAssets) {
            return parent::convert($asset, $basePath);
        }
        $ext = pathinfo($asset, PATHINFO_EXTENSION);
        if(isset($this->commands[$ext])) {
            list($ext, $command) = $this->commands[$ext];
            $from = $basePath . '/' . $asset;
            $to = sprintf('%x', crc32($from . filemtime($from))) . '.' . $ext;
            if($this->forceConvert || @filemtime($am->basePath . '/' . $to) < @filemtime($from)) {
                $this->_runCommand($command, $basePath, $from, $am->basePath . '/' . $to);
            }
            return $am->baseUrl . '/' . $to;

        }
        return $asset;
    }

    protected function _runCommand($command, $basePath, $from, $to) {
        $command = Yii::getAlias($command);
        $command = strtr($command, [
            '{from}' => escapeshellarg($from),
            '{to}' => escapeshellarg($to),
        ]);
        $descriptor = [
            1 => ['pipe', 'w'],
            2 => ['pipe', 'w'],
        ];
        $pipes = [];
        $proc = proc_open($command, $descriptor, $pipes, $basePath);
        $stdout = stream_get_contents($pipes[1]);
        $stderr = stream_get_contents($pipes[2]);
        foreach ($pipes as $pipe) {
            fclose($pipe);
        }
        $status = proc_close($proc);

        if ($status === 0) {
            Yii::trace("Converted $from into $to:\nSTDOUT:\n$stdout\nSTDERR:\n$stderr", __METHOD__);
        } elseif (YII_DEBUG) {
            throw new Exception("AssetConverter command '$command' failed with exit code $status:\nSTDOUT:\n$stdout\nSTDERR:\n$stderr");
        } else {
            Yii::error("AssetConverter command '$command' failed with exit code $status:\nSTDOUT:\n$stdout\nSTDERR:\n$stderr", __METHOD__);
        }

        return $status === 0;
    }

}

```

**Comment by DrDeath72** [comment]
Всплыла проблемка, файлы лежат в css/style.less и img/bla.png и прописаны в less как "../img/bla.png", при конвертировании в корень /assets/ картинки теряются...

**Comment by samdark** [comment]
So solution propose isn't a good idea.

**Comment by DrDeath72** [comment]
ну как вариант обрабатывать css после конвертации и менять пути, но это костыль :)

**Comment by SamMousa** [comment]
@DrDeath72 as far as I can tell, if you use an `AssetBundle` its files will first get published (moved to a public folder).

**Comment by samdark** [comment]
Converting in the target dir is a bad idea because of writes are not atomic. If one is using CDN she may end up with partially done file served.

**Comment by SamMousa** [comment]
@samdark Could we do it in a temp folder and move it from there?
I do agree that you don't want random folders (read: wherever you have a source file that needs compilation) to contain compiled assets.
When locking stuff down it even makes sense to not have write permission on your source folder.

**Comment by samdark** [comment]
Yes. Temp directory such as sub-dir in `runtime` would do.

**Comment by samdark** [comment]
Won't be implemented in 2.0.
